Use verify_guard also to check memory content.

bernhardu opened the pull request:
This should fail at replay time e.g. with not matching register, if the memory returned by of e.g. a ioctl is different than at record time.

bernhardu committed:
Use verify_guard also to check memory content.

This should make a replay fail with mismatched register
if for some reason the memory of e.g. an ioctl is different
than at recording time.

bernhardu commented:
For now just as draft, as I expect at least sioc test to fail, it does at least on my system.

khuey commented:
All tests pass here with this. Kernel 6.13.

bernhardu commented:
> All tests pass here with this. Kernel 6.13.

Thanks for testing. Both system I run have a 6.1.129-1 from Debian.

I added a temporary test to this PR find out if my change to verify_guard does not work, or CI behaves different than my system.

rocallahan commented:
I think it would be good to take this.

rocallahan commented on src/test/util.h:
```diff
@@ -295,8 +295,15 @@ inline static void* allocate_guard(size_t size, char value) {
+  int tmp_errno = errno;
   test_assert(
       memcmp(cp - sizeof(GUARD_VALUE), &GUARD_VALUE, sizeof(GUARD_VALUE)) == 0);
+  static int verify_guard_checksum;
```
Why is this `static`?

bernhardu replied:
This is from a state when I was not yet sure how to get it to the rr register check. I thought first about doing a printf of it before program exit, but that would make it hard to find out when things start to go wrong.
Calling the invalid syscall is now done in each verify_guard, so no need for static.

rocallahan commented on src/test/util.h:
```diff
@@ -295,8 +295,15 @@ inline static void* allocate_guard(size_t size, char value) {
       memcmp(cp - sizeof(GUARD_VALUE), &GUARD_VALUE, sizeof(GUARD_VALUE)) == 0);
+  static int verify_guard_checksum;
+  for (size_t i = 0; i < size; i++)
+      verify_guard_checksum += ((unsigned char*)p)[i];
```
{} around this block.

bernhardu commented:
Thanks for all feedback. I removed the static, added the brackets, and tried to improve the comments.

I still wonder why this did not trigger at khuey's and the CI systems. Are there the names 15 character long, or get this ioctl processed by some software other than the kernel ...

bernhardu commented:
> I still wonder why this did not trigger at khuey's and the CI systems.

Was already in the process to create a Fedora 42 VM, running now with 6.14.0-63.fc42.x86_64.
And found this patch detecting the sioc issue as expected (with a short interface name):
```
benutzer@fedora:~/obj$ bin/rr replay -a sioc-0
...
SIOCGIFTXQLEN(ret:0): ens3 qlen is 1000
[FATAL src/ReplayTask.cc:130:validate_regs()] 
 (task 15551 (rec:15549) at time 277)
 -> Assertion `!comparison.mismatch_count' failed to hold. Mismatched registers, replay vs rec: rdi 0xf43 != 0xee2
...
  real_time:1572.143458 global_time:277, event:`SYSCALL: <unknown-syscall--1>' (state:ENTERING_SYSCALL) tid:15549, ticks:36162
```
